tests/clickhouse/test_query_format.py

```diff
@@ -191,3 +198,104 @@ def test_format_clickhouse_specific_query() -> None:
+
+node_err = IndividualNode(alias="err", data_source=Table("errors_local", ERRORS_SCHEMA))
+node_group = IndividualNode(
+    alias="groups", data_source=Table("groupedmessage_local", GROUPS_SCHEMA)
+)
+node_assignee = IndividualNode(
+    alias="assignee", data_source=Table("groupassignee_local", GROUPS_ASSIGNEE)
+)
+
+TEST_JOIN = [
+    pytest.param(
+        JoinClause(
+            left_node=node_err,
+            right_node=node_group,
+            keys=[
+                JoinCondition(
+                    left=JoinConditionExpression("err", "group_id"),
+                    right=JoinConditionExpression("groups", "id"),
+                )
+            ],
+            join_type=JoinType.INNER,
+            join_modifier=JoinModifier.SEMI,
+        ),
+        SequenceNode(
+            [
+                PaddingNode(None, StringNode("errors_local"), "err"),
+                StringNode("INNER SEMI JOIN"),
+                PaddingNode(None, StringNode("groupedmessage_local"), "groups"),
+                StringNode("ON"),
+                SequenceNode([StringNode("err.group_id=groups.id")]),
+            ]
+        ),
+        (
+            "errors_local err INNER SEMI JOIN groupedmessage_local groups "
+            "ON err.group_id=groups.id"
+        ),
+        id="Simple join",
+    ),
+    pytest.param(
+        JoinClause(
+            left_node=JoinClause(
+                left_node=node_err,
+                right_node=node_group,
+                keys=[
+                    JoinCondition(
+                        left=JoinConditionExpression("err", "group_id"),
+                        right=JoinConditionExpression("groups", "id"),
+                    )
+                ],
+                join_type=JoinType.INNER,
+                join_modifier=JoinModifier.SEMI,
+            ),
+            right_node=node_assignee,
+            keys=[
+                JoinCondition(
+                    left=JoinConditionExpression("err", "group_id"),
+                    right=JoinConditionExpression("assignee", "id"),
+                )
+            ],
+            join_type=JoinType.INNER,
+        ),
+        SequenceNode(
+            [
+                SequenceNode(
+                    [
+                        PaddingNode(None, StringNode("errors_local"), "err"),
+                        StringNode("INNER SEMI JOIN"),
+                        PaddingNode(None, StringNode("groupedmessage_local"), "groups"),
+                        StringNode("ON"),
+                        SequenceNode([StringNode("err.group_id=groups.id")]),
+                    ]
+                ),
+                StringNode("INNER JOIN"),
+                PaddingNode(None, StringNode("groupassignee_local"), "assignee"),
+                StringNode("ON"),
+                SequenceNode([StringNode("err.group_id=assignee.id")]),
+            ]
+        ),
+        (
+            "errors_local err INNER SEMI JOIN groupedmessage_local groups "
```

Comment:
Should there not be some brackets around this? Or is that done somewhere else?

Reply:
Actually no. Clickhouse does not want parentheses around a join expression. It does not work if I add them. I thought it would as well.